src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,134 @@
+      this.trieMap(normK.substring(0, 1)) = next
+      this
+    }
+  }
+
+  def get(letter:String): Trie = trieMap.getOrElse(letter, null)
+
+  def putAll(map: Map[String, String]): Trie = {
+    map.foreach {case (k, v) => {this.put(k, v)}}
+    this
+  }
+
+  def checkForWord(normText:String, index: Int): (Int, String) = {
+      var current = this.get(normText(index).toString)
+      var count = 1
+      var longestWord = ""
+      var longestWordKeyLength = 0
+      while (current != null && index + count != normText.length()) {
+        if (!current.value.isEmpty) {
+          longestWord = current.value
+          longestWordKeyLength = count
+        }
+        current = current.get(normText(index + count).toString)
+        count += 1
+      }
+
+      val foundWord = longestWordKeyLength > 0
+      if (foundWord) {
+        while (index + longestWordKeyLength < normText.length() &&
+          (normText(index + longestWordKeyLength).toString() matches """[a-zA-Z0-9_]""")) {
+          longestWordKeyLength += 1
+        }
+      }
+
+      (max(index + 1, index + longestWordKeyLength), longestWord)
+    }
+
+  def mapText(inputText:String): String = {
+    val normText: String = normFunction(inputText)
+    var outputText: StringBuilder = new StringBuilder("")
+    var i = 0
+    do {
+      val (newIndex, word) = checkForWord(normText, i)
+      if (word.isEmpty()) {
+        outputText ++= normText.substring(i, newIndex)
+      } else {
+        outputText ++= word
+      }
+      i = newIndex
+    } while (i != normText.length())
+    outputText.toString()
+  }
+
+  override def toString(): String = "Val:" + this.value + " Rest:\n" + trieMap.toString()
+}
+
+object TextPreprocessor extends ComplexParamsReadable[TextPreprocessor]
+
+/** <code>TextPreprocessor</code> takes a dataframe and a dictionary
+  * that maps (text -> replacement text), scans each cell in the input col
+  * and replaces all substring matches with the corresponding value.
+  * Priority is given to longer keys and left to right.
+  */
+
+class TextPreprocessor(val uid: String) extends Transformer
+  with HasInputCol with HasOutputCol with Wrappable with ComplexParamsWritable {
+  def this() = this(Identifiable.randomUID("TextPreprocessor"))
+
+  val map = new MapParam[String, String](this, "map", "Map of substring match to replacement")
+
+  /** @group getParam */
+  def getMap: Map[String, String] = get(map).getOrElse(Map())
+
+  /** @group setParam */
+  def setMap(value: Map[String, String]): this.type = set("map", value)
+
+  /** @param dataset - The input dataset, to be transformed
+    * @return The DataFrame that results from column selection
+    */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val spark = dataset.sparkSession
+    val inputIndex = dataset.columns.indexOf(getInputCol)
+    val trie = new Trie().putAll(getMap)
+    val broadcastedTrie = spark.sparkContext.broadcast(trie)
+
+    if (inputIndex == -1)
```

Comment:
more idiomatic to use `require` when throwing `IllegalArgumentExceptions`

Reply:
Require for what exactly? Can you link me to the require you are referring to?

Reply:
it's in the Predef, which is already imported in all scala scopes

https://github.com/scala/scala/blob/2.11.x/src/library/scala/Predef.scala#L222

Reply:
Thanks! I added the change. Much cleaner now.